### ISSUE TITLE
Specify ClusterAutoscaler's priority in shoot CR

### ISSUE THREAD

**Issue by dongyingbo** [gardener/gardener#10683]
**How to categorize this issue?**

/area auto-scaling
/kind enhancement

**What would you like to be added**:
Currently ClusterAutoscaler's priority configmap cluster-autoscaler-priority-expander is installed by user after cluster creation.
It should be part of shoot CR. 

**Why is this needed**:
We have more than one node groups to allow system components for cluster resilience.
But the cheapest one should take priority for TCO.
But we can not do that because configmap cluster-autoscaler-priority-expander is created by us after cluster ready.
I would like to make the configmap part of shoot yaml, so that gardener controller can create it in between control plane ready and system component deployment.

**Comment by sunir1** [comment]
I support this request, and would add that it is not only for TCO, but also for performance.
We use fallback options (by priority), due to capacity issues by hyper-scalers which are out of our control.
And ideally we would like the 1st/preferred worker-group to be used for better performance and/or cost.

Having randomness (while capacity is available), is limiting us.

**Comment by vlerenc** [comment]
I am experiencing the same problem. Pools are defined in the shoot spec, but priorities in some other resource that can be created/modified only after the cluster was created (and never in one go/step). That makes it hard to change both in combination, but since the expander configuration depends (regex-matches) on the pools, that is somewhat unfortunate. They actually "belong together" /are two sides of the same coin.

Maybe Gardener could pass through the expander configuration via an extension, so that the CA configmap doesn't have to become (worker pool) API? If CA is later replaced (hypothetically) by another CA/recommender or if MCM does it natively (hypothetically), there would be no harm and whatever the new CA/recommender needs can then be passed on by another extension (or it becomes part of the worker pool API if MCM deals with it natively).

cc @unmarshall FYI

**Comment by vlerenc** [comment]
@dongyingbo I didn't say that, actually the opposite. Because CA is only one solution (even though the industry/community standard) and the configuration is specific to it, a CA extension would be a more appropriate way to integrate it into Gardener (as one of many options), not polluting the shoot API - and then of course the configuration, in the shoot, for the CA extension, would be the native one.

**Comment by timuthy** [comment]
Also see https://github.com/gardener/gardener/issues/10765 for @vlerenc's extension vote.

**Comment by timuthy** [comment]
@vlerenc and I had an internal discussion about the extension topic. We agreed that factoring out CA into an extension, and thus allow other autoscaler solutions, might make sense on the long run (approx. end of 2025).

For the time begin, we plan to continue with the in-tree deployment routine which we can enhance by an Priority Expander configuration, passed through the Shoot API:

```
spec:
  kubernetes:
    clusterAutoscaler:
      expanderConfig:
        configMapName:
```

**Comment by vlerenc** [comment]
@timuthy Will it be really `configMapName` or will inline priorities be also supported? BTW, in case you plan only for a config map name (less convenient than inline), please note that the audit log config is called [`configMapRef.name`](https://github.com/gardener/gardener/blob/master/example/90-shoot.yaml#L189-L192).

**Comment by rfranzke** [comment]
I don't think we will support inline priorities to not overload the `Shoot` resource. I guess this can grow quite large if you have many worker pools.

> please note that the audit log config is called [`configMapRef.name`](https://github.com/gardener/gardener/blob/master/example/90-shoot.yaml#L189-L192).

I think there are already some inconsistencies (we have other fields called `secretName` or similar, i.e., without a `ref`). This should be harmonized when a new API version is introduced.

**Comment by dongyingbo** [comment]
May you clarify that where and when contents of the configmap comes from if using configMapName/configMapRef.name?

**Comment by tobschli** [comment]
The `ConfigMap` should be placed in the project's namespace where the `Shoot` will be created.
Before the deployment of the CA, it will be placed in the right spot for the CA to use.

**Comment by dongyingbo** [comment]
> The `ConfigMap` should be placed in the project's namespace where the `Shoot` will be created. Before the deployment of the CA, it will be placed in the right spot for the CA to use.

Am I correct that,
- It will be created by user who creates Shoot CR
- It should be created before Shoot CR as it is the only way to ensure it is before CA from Gardener user side
- If there are multiple shoots in project, user needs to ensure CM name unique among Shoots. Then there is might be problem for dynamic created dev shoots. Perhaps it will end up by use same configmap in one project.

**Comment by tobschli** [comment]
> It will be created by user who creates Shoot CR
It does not need to be the same user, but some user in the project. Of course the configuration within this `ConfigMap` depends on the worker groups of the Shoot. So the user who creates the `Shoot`, who also wants to consume such a configuration, needs to make sure this configuration works for the `Shoot` they are trying to create.

> It should be created before Shoot CR as it is the only way to ensure it is before CA from Gardener user side
Yes

> If there are multiple shoots in project, user needs to ensure CM name unique among Shoots. Then there might be problem for dynamic created dev shoots. Perhaps it will end up by using same configmap for one project.
I am not quite sure if I understand this correctly, correct me if I am wrong.
You do not have to use a single `ConfigMap` per project.
If you have `Shoot`s with different worker groups, i.e. the worker group for `shoot-1` has a machine type, that `shoot-2` does not have, you cannot use the same priority config.
In this case you create two `ConfigMap`s that are tailored for the specific worker group configurations and reference them in the correct `Shoot` respectively.

**Comment by vlerenc** [comment]
> I don't think we will support inline priorities to not overload the Shoot resource. I guess this can grow quite large if you have many worker pools.

@rfranzke Worker pools can become very large, but priorities will never be large/can never become large (by definition, because they only list names and cannot include anything else).

The problem of not allowing inline priorities is that I then need to build a lot of logic around updating the worker pools and their priorities.

In my eyes, they are **one unit** and it makes no sense to separate them (separating them is just an artefact of MCM and CA, but they only makle sense to be updated jointly, because priorities refer to pools - always).

So, if inlining is not supported, I must manage config maps myself. I probably shouldn't update the currently in use priorities config map, because if it gets reconciled (because a subsequent workers update fails and/or someone else triggers a shoot reconciliation before the workers section gets updated), it wouldn't fit anymore the current workers section. So, I would have to create a new config map, refer to that in the shoot spec when I also update the workers section, wait until successful reconciliation, and then garbage-collect/delete the old one. That seems a bit complex/something I would rather like to avoid (especially in the context of CI/CD/git-ops triggered operations like landscape-setup).

I would have clearly preferred inlining, as a second option, e.g. `name` can refer to a config map, but `data` can bring the expander config map content inline. Would that be ok for you as well @rfranzke (cc @tobschli)?

**Comment by dongyingbo** [comment]
>I am not quite sure if I understand this correctly, correct me if I am wrong.
You do not have to use a single ConfigMap per project.
If you have Shoots with different worker groups, i.e. the worker group for shoot-1 has a machine type, that shoot-2 does not have, you cannot use the same priority config.
In this case you create two ConfigMaps that are tailored for the specific worker group configurations and reference them in the correct Shoot respectively.

@tobschli 
As you know we have shoots for different hyperscaler in same develop/test project. And their workergroups are definitely different.
Actually even for shoots of same hyperscaler, workergroups are also different.

We used to have configmap:shoots/project as 1:1.
But now it seems impossible or hard to achieve.
You mean it is not necessary to make it 1:all, and we can make it 1:m.

I am not sure if it flexible enough for us. I am in the impression that it is not, or an downgrade from flexibility perspective at least.

**Comment by vlerenc** [comment]
@dongyingbo:
> I would like to make the configmap part of shoot yaml, so that gardener controller can create it in between control plane ready and system component deployment.

You actually need it to be available while the control plane is in creation, i.e. right before CA and MCM act, i.e. decide about a machine and add it.

You can steer that via `minimum` number of nodes - that's what we do. Since Gardener scales from zero, that works nicely, even if other machine types become necessary.

But overall, it seems a slim time window to cause any trouble. Is that really the motivation or is it convenience, because that's my reason: Pools and priorities belong together, hence I would like to control them in one place, the `shoot` spec. Since I already have a controller that manages clusters, I like to leave control there and not build logic to reach the cluster and modify a `configmap` there. Plus, sometimes I cannot reach that cluster, because it's behind a firewall. Gardener overcomes this problem elegantly by shoots talking to seeds and them talking in turn to Gardener, so only this way I am guaranteed to get stuff into the cluster or to the CA. Hence, the proposal is a huge improvement for me, because now I can manage the `configmap` in the garden cluster instead.

Still, even better would be to configure the pools and the priorities without this `configmap` detour, especially because priorities refer to pools, so they form a "unit" and I would welcome it very much, if they can be configured in the `shoot` spec directly @tobschli.

But the time window where a bad decision could be made by CA can be bridged by using the `minimum` configuration, which is what it's goof for.

> We used to have configmap:shoots/project as 1:1. But now it seems impossible or hard to achieve. You mean it is not necessary to make it 1:all, and we can make it 1:m.

I don't think, this is true. You can stick to the current approach, but then you have not improved anything, true.

You can use the current proposal and decide for yourself, whether you map 1:1 (each cluster gets it's own `configmap`), 1:N (all clusters get the same `configmap`), or N:M (some clusters share their `configmap`), but in all cases you need to manage the `configmaps`, true (that's what I would rather like to avoid, but it's no neck-breaker).

> I am not sure if it flexible enough for us yet. I have an intuition that it is not, or a downgrade from flexibility perspective at least.

That I cannot see. The PR describes a more flexible solution - you can now do things that weren't possible before (hence, more flexible).

@sunir1 
> I support this request, and would add that it is not only for TCO, but also for performance. We use fallback options (by priority), due to capacity issues by hyper-scalers which are out of our control. And ideally we would like the 1st/preferred worker-group to be used for better performance and/or cost.
> Having randomness (while capacity is available), is limiting us.

Same thing as above, there is no real randomness. The time window between the creation of a cluster and applying the `configmap` can be very small/irrelevant, especially for this use case. Productive clusters (in contrast to test clusters) are probably not created in so huge numbers, that a few seconds in between their creation and writing and applying the `configmap` becomes relevant due to a cloud provider outage or being out of stock.

Even if, you have `minimum`, taints and tolerations, and scale-from-zero to work with. That's what we do.

---

To sum it up, the current PR is already a huge improvement, I think, because we can now manipulate the priorities inside the garden cluster where also the `shoot` resource resides (solves the network reachability problem and even eliminates the time window problem where CA has no `configmap` yet, but that was never a real neck-breaker, I think - if it is, please share how), but I would love if this would be even more conveniently possible right inside the `shoot` spec without me being forced to managed a bunch of `configmaps` for something (priorities) that appear to be so tightly linked to the pools that sit in the `shoot` spec, please.

**Comment by rfranzke** [comment]
I think @tobschli and I have misunderstood the requirements. We have clarified them once again in a call with @vlerenc. As a result, this is the new proposal:

- It is not only about initially creating the CA `ConfigMap`. Instead, with managing the CA priorities in the `Shoot` spec, the approach of manipulating the CA `ConfigMap` in the shoot cluster's `kube-system` namespace becomes obsolete. All changes to the priorities must go via the `Shoot` spec and are also reconciled beyond day 1.
- Inlining the priorities makes more sense compared to referencing them in a `ConfigMap` since they strongly belong to the worker pools in the `Shoot` spec (i.e., when the pools are changed, the priorities also might be changed).

@dongyingbo Thoughts? Does this also match with your expectation?

@tobschli I guess it makes sense to close https://github.com/gardener/gardener/pull/10880 and open a new PR implementing the above.

**Comment by rfranzke** [comment]
> How I understand it now is that we add a field to the `Shoot.spec.provider.workers[]`, right? I imagine something like `priority`, that if specified for one Worker group:
> 
> * Either needs to be validated that all worker groups have this specified
> * Some kind of defaulting takes place
> 
> From this point, the expander mode of the CA needs to be set to `priority`, as well as the `ConfigMap` needs to be generated and placed onto the respective `Shoot` accordingly.

No, I don't think (correct me if I'm wrong) that it makes sense to specify a priority for each worker pool and then let `gardenlet` craft the priority` configuration for CA.
Instead, let's just let the user inline the priority configuration in the `clusterAutoscaler` section (similar to now, just w/o `ConfigMap`).

**Comment by tobschli** [comment]
That is also fine by me :)
I just thought, as this configuration strongly belongs to the worker pools, we could directly combine this.

One benefit of the worker-pool approach could be that it would become impossible to create an invalid configuration.

Although it can be confusing for a `Shoot` owner to maintain the CA priorities at the worker pool configuration.

**Comment by rfranzke** [comment]
@vlerenc Can you give some input to this? As far as I understood, you don't have to necessarily define a priority per pool (since it could be handled with `*` wildcards), or?

**Comment by vlerenc** [comment]
@rfranzke @tobschli Wow... that is a difficult question. First the easy answer: Yes, it is not necessary to even define a `configmap` for CA (as you know). CA will already try to pick the machine type (node group) with the least waste, but sometimes you want to steer that - then you need it, e.g. if you have made certain discount/reserved instances purchases - that's nothing CA can know. And indeed, then you might work with a regex to catch those, e.g. prefer `m5.*` and avoid `m4.*` or `r5.*` that you may have as fallback machine types. Finally, there is even the best practice, to close any such `configmap` with a catch-all `.*`, see https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/expander/priority/readme.md#configuration.

However, I am sure, we'll find a way to deal with this. No priority means catch-all priority of 0. If you want the same priorities for multiple worker pools (node groups in CA speak), then you give them the same priority and since the `configmap` is actually a list (of priorities) of list (of regex expressions), that can be mapped into the CA world easily. No show-stopper.

So, the difficult question is rather, how deep you want to integrate "priorities" into our shoot API. Right now, we think only in terms of CA and since it doesn't offer much here, it would appear reasonable to put priorities right into the worker pool definition (having the list of priorities with list of regex expressions isn't yielding any additional advantages). But can it be, that this will bite us later? Can it be that we later find cases where priorities must be defined at a higher level - independent of individual worker pools? I don't know. Do you @unmarshall?

Also, when I discussed the CA -> extension topic with @timuthy, we agreed that any overhaul is probably time-consuming, so we thought in the end, let's do the simplest solution for now (new/additional CA field in the shoot spec), because we have introduced the CA API already into the shoot API and the moment we drop CA and start using our own node autoscaling recommender, we have to clean that up anyway.

So, I am fine with both approaches, but my thoughts are:
- Having the priority right with the pool feels a lot more "natural" and will make this feature also easier to master for our end users that then don't have to concern themselves with CA and the format of that (proprietary) `configmap`.
- If we do not find issues with this approach, this API won't have to be changed/can remain as is even.
- If we do find issues with this approach, e.g. when we introduce our own node autoscaling recommender (or another one), we must change/extend the API, but we would have to do that anyway if we go with the simplest solution (as a dedicated CA field in the shoot spec).
So, I am more in favour of the new proposal by @tobschli. He came up with it, free of the CA baggage that we all have.
It would be more natural and in the best/better case has a chance of remaining stable API also in the future.

**Comment by vlerenc** [comment]
@unmarshall called me (independently and by chance) and he is also in favour of abandoning the CA API/`configmap` and having the `priorities` right with the pools as @tobschli suggested.

He mentioned also Karpenter - they call it `weights` btw, so maybe that's the better term?

See example (although Karpenter works again slightly differently, but that's another story/not for here): https://github.com/aws/karpenter-provider-aws/blob/main/designs/provisioner-priority.md?plain=1#L79-L110

In general, the `configmap` is a hack by CA, like it's also a hack that CA "inspects" existing nodes and guesses that more such nodes with come up with the same characteristics (this also made scale-from-zero so hard for us to implement with CA). CA is meant to be deployed into the cluster it auto-scales and as such, a config map with regex expressions makes sense for it, but in our case, we do have a source of truth where the worker pools (=node groups) are cleanly defined, our shoot spec. So, it's more natural to define the `priorities`/`weights` there instead of forcing the users to know that we deploy CA (implementation detail) and that we map multi-zonal worker pools to zonal node groups 1:N (implementation detail), that they then have to refer to with regex expressions (matching our generated node group names).

So, `priorities`/`weights` right with the individual pools is then the decision (no `configmap`). It comes with the most advantages.

**Comment by tobschli** [comment]
I would be in favour of calling the field `priority`, as it, by today, specifies the priority in CA.
When looking in the future of the aforementioned autoscaling extension, I can imagine that this field will need some rework (e.g. going from `.spec.provider.workers[].priority` to something like `.spec.provider.workers[].scaling.priority/weight` or something completely provider specific).
This is nothing but a guess though 😄

**Comment by elankath** [comment]
Thanks for considering this. The CA priority map is definitely error-prone to define and many people get it wrong as they make elementary mistakes in NG names and regular expressions and require support. Of-course there exists the hard job of convincing folks who are used to the CA and already coded so many priority expander config maps.

**Comment by tobschli** [comment]
As far as I see this at the moment, there also needs to be an adaption in the provider Extensions if we want to support this.
The configuration of the nodeGroups that CA can scale are based on the names of `MachineDeployment`s and not the Worker configuration of the `Shoot` itself. ([ref](https://github.com/gardener/gardener/blob/master/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go#L482)).
There is also no mapping between Worker group configuration and `MachineDeployment`s that we could access.
The relation between Worker configuration and `MachineDeployment`s also is not `1:1` (e.g. for HA, there is a `MachineDeployment` for each zone) and the names of these are generated in the provider extensions respectively.

Although it seems that the extensions include the worker pool name in the `MachineDeployment` name ([ref](https://github.com/gardener/gardener-extension-provider-aws/blob/master/pkg/controller/worker/machines.go#L235)).
As the CA priority expander configuration allows regex, we _coul_d determine the `MachineDeployment`s for a Worker pool that way.
Although I would not prefer to do it this way, as extensions do not _have_ to include the pool name in the `MachineDeployment`. This would also imply an implicit contract that extensions have to name the `MachineDeployment`s a certain way

Therefore I would go on with implementing the change in the `MachineDeployment` API and implement it that way.

**Comment by dongyingbo** [comment]
@vlerenc 
> You actually need it to be available while the control plane is in creation, i.e. right before CA and MCM act, i.e. decide about a machine and add it.  

Yes. I need it be available right before CA and MCM act.
I say "after control plane is ready" because I assumed it would be installed into shoot cluster's kube-system namespace still instead of into garden cluster.  


> You can steer that via minimum number of nodes - that's what we do. Since Gardener scales from zero, that works nicely, even if other machine types become necessary.  

I do not quite understand here.
My motivation is:
We have WGs allowing system components with priority values default-arm>default>default-f1. We want default-arm is used for gardener system components assuming there are enough available VMs on hyperscaler side.  
The current situation is that the 3 WGs share same possibility as their sizes are same.  

How can minimum number of nodes property help here? We have minimum 0 for all the 3 WGs now.  

> now I can manage the configmap in the garden cluster instead  

AFAIK, our operation colleagues modify the configmap using kubectl directly for emergency.  
But they will have access both to shoot cluster and garden cluster for emergency. So it should be OK for them to move the priority configuration location. Only they need to be aware of the change.

**Comment by dongyingbo** [comment]
> I think [@tobschli](https://github.com/tobschli) and I have misunderstood the requirements. We have clarified them once again in a call with [@vlerenc](https://github.com/vlerenc). As a result, this is the new proposal:
> 
> * It is not only about initially creating the CA `ConfigMap`. Instead, with managing the CA priorities in the `Shoot` spec, the approach of manipulating the CA `ConfigMap` in the shoot cluster's `kube-system` namespace becomes obsolete. All changes to the priorities must go via the `Shoot` spec and are also reconciled beyond day 1.
> * Inlining the priorities makes more sense compared to referencing them in a `ConfigMap` since they strongly belong to the worker pools in the `Shoot` spec (i.e., when the pools are changed, the priorities also might be changed).
> 
> [@dongyingbo](https://github.com/dongyingbo) Thoughts? Does this also match with your expectation?
> 
> [@tobschli](https://github.com/tobschli) I guess it makes sense to close [#10880](https://github.com/gardener/gardener/pull/10880) and open a new PR implementing the above.

Hi @rfranzke ,
Thanks, yes it meets both my expections:
- CA `ConfigMap` gets created initally
- Priorities is inlined in the `Shoot` spec

**Comment by vlerenc** [comment]
@dongyingbo I cannot follow. Here some comments:
- The CA is part of the control plane. Whether a configmap exists or not.
- I still do not understand, why a few seconds matter (the time you need to install the configmap before more workload is created), but we anyway plan now to make the priorities configurable per shoot resource and so it probably doesn't matter whether I understand it or not, but I don't.
- Naturally, the number of (existing) nodes matters (if there is capacity in the cluster, because of existing nodes, CA doesn't get involved) and if all worker pools have a minimum of 0 and you have no taints either, CA will pick the machines with least waste. So, you did nothing of the above?
E.g. you could set the minimum to 3 nodes (or whatever is the minimum is in your case) for the ARM type to get started, and you could taint the other machine types, if you really want certain system pods to go only to a particular worker pool.

**Comment by dongyingbo** [comment]
> [@unmarshall](https://github.com/unmarshall) called me (independently and by chance) and he is also in favour of abandoning the CA API/`configmap` and having the `priorities` right with the pools as [@tobschli](https://github.com/tobschli) suggested.
> 
> He mentioned also Karpenter - they call it `weights` btw, so maybe that's the better term?
> 
> See example (although Karpenter works again slightly differently, but that's another story/not for here): https://github.com/aws/karpenter-provider-aws/blob/main/designs/provisioner-priority.md?plain=1#L79-L110
> 
> In general, the `configmap` is a hack by CA, like it's also a hack that CA "inspects" existing nodes and guesses that more such nodes with come up with the same characteristics (this also made scale-from-zero so hard for us to implement with CA). CA is meant to be deployed into the cluster it auto-scales and as such, a config map with regex expressions makes sense for it, but in our case, we do have a source of truth where the worker pools (=node groups) are cleanly defined, our shoot spec. So, it's more natural to define the `priorities`/`weights` there instead of forcing the users to know that we deploy CA (implementation detail) and that we map multi-zonal worker pools to zonal node groups 1:N (implementation detail), that they then have to refer to with regex expressions (matching our generated node group names).
> 
> So, `priorities`/`weights` right with the individual pools is then the decision (no `configmap`). It comes with the most advantages.

Currently we define 3 workergroups default-arm/default/default-f1 for same workload. And the only differences for them is machine type.
It will be much better if we can,
- define a machine type list for one workergroup
- machine types are selected as same as "priority,least-waste" expander

**Comment by dongyingbo** [comment]
@vlerenc 
First of all I would like to clarify that I am talking about temporary clusters for dev/ci-cd but not long-living clusters.
And it makes sense to us because we create many dev/ci-cd clusters everyday and would like to save some cost.

The diagram shows how I understand,
![Image](https://github.com/user-attachments/assets/698d934f-1ee8-4d60-af26-a659fbbc5d5f)
The comment in the diagram shows where/when the problem happens.
At the time, CA priority does not work yet. And CA works with least-waste expander.
In this way VMs are provisioned randomly from the 3 WGs as their machine sizes are same.
We set their minum to 0. We can not set them no-zero because any of them might be out of capacity on hyperscaler side and it is why we create 3 WGs.

**Comment by vlerenc** [comment]
@dongyingbo You can still set the minimum to N - it's unlikely that the hyperscaler if often out of capacity. That is a corner case. If it is, CA will be involved and provision the other machines. You can also create the configmap the moment the KAPI comes to live. That's why I said seconds, because the KAPI comes before the nodes.

But either way, this ticket will allow to set the priorities right inside the shoot resource, so what I say doesn't matter too much. I only wanted to express that your problem can be mitigated. We still want the priorities in the shoot resource, because they are so closely related to the worker pool definition. So, all is good, we plan to add it.

**Comment by tobschli** [comment]
With the proposed changes, the issue can be resolved, as priorities can be specified like:
```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
spec:
  provider:
    workers:
    - name: important-worker-group
      minimum: 1
      maximum: 10
      maxSurge: 1
      maxUnavailable: 0
      priority: 100
```

During the deployment of the CA, the priority-expander `ConfigMap` will be generated and deployed.

Another approach, with the same result, would be to adapt the CA to directly take the priority information from the `MachineDeployment`s. Downside would be higher maintenance in our fork.

**Comment by dongyingbo** [comment]
> With the proposed changes, the issue can be resolved, as priorities can be specified like:
> 
> apiVersion: core.gardener.cloud/v1beta1
> kind: Shoot
> spec:
>   provider:
>     workers:
>     - name: important-worker-group
>       minimum: 1
>       maximum: 10
>       maxSurge: 1
>       maxUnavailable: 0
>       priority: 100
> During the deployment of the CA, the priority-expander `ConfigMap` will be generated and deployed.
> 
> Another approach, with the same result, would be to adapt the CA to directly take the priority information from the `MachineDeployment`s. Downside would be higher maintenance in our fork.

Hi @tobschli ,
We have many WGs, more than 80 in one shoot for example. And there are more than 60 lines for its priorities setting.
Placing priority into WG definition may lead to harder for priority overview and debug IMHO.
Of course we can check the generated configmap but it is incontinent during development.
What is your opinion please?

Do you think it is worth to consider support machine type list for WG like @vlerenc 's [example](https://github.com/aws/karpenter-provider-aws/blob/main/designs/provisioner-priority.md?plain=1#L79-L110) above?
In this way Gardener can hide priority from users and make users life easier IMHO.

**Comment by dongyingbo** [comment]
> [@dongyingbo](https://github.com/dongyingbo) You can still set the minimum to N - it's unlikely that the hyperscaler if often out of capacity. That is a corner case. If it is, CA will be involved and provision the other machines. You can also create the configmap the moment the KAPI comes to live. That's why I said seconds, because the KAPI comes before the nodes.
> 
> But either way, this ticket will allow to set the priorities right inside the shoot resource, so what I say doesn't matter too much. I only wanted to express that your problem can be mitigated. We still want the priorities in the shoot resource, because they are so closely related to the worker pool definition. So, all is good, we plan to add it.

Hi @vlerenc 
Thanks for your patient.
Just to provide some background here.
> it's unlikely that the hyperscaler if often out of capacity

Actually it is not so rare to see hyperscaler is out of capacity. And we have done many on our side to fight against it for cluster resilience.
For example we reported https://github.com/gardener/machine-controller-manager/issues/952. And we saw resource scarcity almost every day during that.

**Comment by vlerenc** [comment]
> Placing priority into WG definition may lead to harder for priority overview and debug IMHO.

It is now where it belongs, where you manage zones, machine types, images, etc. So, we argue the other way.

Besides, if you have 80 or more worker pools, you hopefully generate them, but either way (manual or automatic), all aspects but priorities where there and now all aspects are there.

> Of course we can check the generated configmap but it is incontinent during development.

Why would you check it, if it works? We generate thousands of things for you.

> Do you think it is worth to consider support machine type list for WG like @vlerenc 's [example](https://github.com/aws/karpenter-provider-aws/blob/main/designs/provisioner-priority.md?plain=1#L79-L110) above?

This is from Karpenter, a solution from Amazon covering AWS. The majority (including Gardener) use the Cluster-Autoscaler.

> In this way Gardener can hide priority from users and make users life easier IMHO.

I do not understand. On the one hand side you opened a ticket in which you wanted to handle priorities differently and now you suggest to hide them, so that "life gets easier"? If the latter is your goal, don't use priorities. But if you use them, you have to deal with them, understand them, set and steer them where you want them to be.

> Actually it is not so rare to see hyperscaler is out of capacity. And we have done many on our side to fight against it for cluster resilience.

Of course they are often out of resources, but it's not all that relevant for a time window of seconds. You set `minimum` of your preferred worker pool to what you require as a minimum and if they are out of stock, CA/MCM will give you the other machines (if you manage to automatically set the configmap within a few seconds, then even according to your priorities).

But this doesn't matter now anymore because of @tobschli PR.

**Comment by rfranzke** [comment]
Current status: Waiting for https://github.com/gardener/gardener-extension-provider-alicloud to bump their `github.com/gardener/gardener` dependency to at least `v1.114`

**Comment by tobschli** [comment]
I just opened the last PR for our well-known provider-extensions.
After it is merged and released, I will go ahead and remove the best-effort approach, completing the integration of it.

**Comment by rfranzke** [comment]
All provider extension PRs have been merged (partly not released).

Next step: Cleanup https://github.com/tobschli/gardener/blob/0df3391737ca7c36cd58279845a46c832ce4b422/pkg/component/autoscaling/clusterautoscaler/cluster_autoscaler.go#L683-L685 after Gardener v1.126 was released.
